utils_nlp/models/bert/question_answering.py

```diff
@@ -0,0 +1,266 @@
+
+        # Prepare optimizer and schedule (linear warmup and decay)
+        no_decay = ["bias", "LayerNorm.weight"]
+        optimizer_grouped_parameters = [
+            {
+                "params": [
+                    p
+                    for n, p in self.model.named_parameters()
+                    if not any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.01,
+            },
+            {
+                "params": [
+                    p for n, p in self.model.named_parameters() if any(nd in n for nd in no_decay)
+                ],
+                "weight_decay": 0.0,
+            },
+        ]
+        optimizer = AdamW(optimizer_grouped_parameters, lr=learning_rate, eps=1e-8)
+
+        if warmup_proportion:
+            warmup_steps = t_total * warmup_proportion
+        else:
+            warmup_steps = 0
+
+        scheduler = WarmupLinearSchedule(optimizer, warmup_steps=warmup_steps, t_total=t_total)
+
+        global_step = 0
+        tr_loss = 0.0
+        self.model.zero_grad()
+        train_iterator = trange(int(num_epochs), desc="Epoch")
+        for _ in train_iterator:
+            for batch in tqdm(train_dataloader, desc="Iteration", mininterval=60):
+                self.model.train()
+                batch = tuple(t.to(device) for t in batch)
+                inputs = {
+                    "input_ids": batch[0],
+                    "attention_mask": batch[1],
+                    "token_type_ids": batch[2],
+                    "start_positions": batch[3],
+                    "end_positions": batch[4],
+                }
+
+                outputs = self.model(**inputs)
+                loss = outputs[0]  # model outputs are always tuple in pytorch-transformers
+
+                loss = (
+                    loss.mean()
+                )  # mean() to average on multi-gpu parallel (not distributed) training
+
+                loss.backward()
+                torch.nn.utils.clip_grad_norm_(self.model.parameters(), max_grad_norm)
+
+                tr_loss += loss.item()
+
+                scheduler.step()  # Update learning rate schedule
+                optimizer.step()
+                self.model.zero_grad()
+                global_step += 1
+
+        logger.info(" global_step = %s, average loss = %s", global_step, tr_loss / global_step)
+
+        if cache_model:
+            if not os.path.exists(self.cache_dir):
+                os.makedirs(self.cache_dir)
+
+            logger.info("Saving model checkpoint to %s", self.cache_dir)
+            # Save a trained model, configuration and tokenizer using `save_pretrained()`.
+            # They can then be reloaded using `from_pretrained()`
+            model_to_save = (
+                self.model.module if hasattr(self.model, "module") else self.model
+            )  # Take care of distributed/parallel training
+            model_to_save.save_pretrained(self.cache_dir)
```

Comment:
The model file name and configuration file name are internal constants. See https://github.com/huggingface/pytorch-transformers/blob/master/pytorch_transformers/modeling_utils.py. save_pretrained and from_pretrained both use these constants. 
I made some changes so that the user can choose if they want to overwrite existing models. Please take a look.